### ISSUE TITLE
NodeJS version?

### ISSUE THREAD

**Issue by dmitrypisanko** [transcranial/keras-js#32]
Do you planning to adapt this module for nodeJS?

Thanks

**Comment by unsalted** [comment]
I haven't tested it but it should work in Node, you just won't be able to use webgl, and even then you may be able to add a webgl context to Node through something like: https://github.com/stackgl/headless-gl

That said... why?  It's faster to run keras/tensorflow directly?

**Comment by dmitrypisanko** [comment]
I try, but script exit with errors
```
undefined:6
var canvas = document.createElement('canvas');
             ^
ReferenceError: document is not defined
```
I am allready train my model and have HDF5 weights file. I just want use them in my nodeJS backend. I can try adapt keras-js, but decided to ask first :)

**Comment by unsalted** [comment]
Yeah I just took a look, it's a little silly that it expects the dom to be present, in an ideal world this would separate itself from xhr requests, the dom, and the gpu and hand that off to the dev (because we are all opinionated about how we like to manage those).  Apposed to making assumptions about usage. The only reason it asks for canvas is because it's checking for a webgl texture size which you could argue should not the job of this lib, but should be the job of the dev who implements it.

Not to mention the expectation that weblas is globally scoped is also a little frustrating considering that most people are using some variant of AMD at this point.

https://github.com/transcranial/keras-js/blob/29e56abd9aaf934f6f8c7e1d6e7659da85cfd858/src/constants.js

Griping aside it's an impressive feat that it works as well as it does.

**Comment by dmitrypisanko** [comment]
So i'll have to wait for nodeJS version or try to do it by myself? :)

**Comment by transcranial** [comment]
I've tried it -- can definitely make this work in nodejs after stripping away weblas/canvas/webgl. What's your compelling reason for wanting to run it in nodejs, though?

**Comment by unsalted** [comment]
In theory you could actually get webgl and canvas working in Node, which would make a compelling case for serving (lightweight) model predictions or parsing data in a single code base app.
But this would require injecting those dependencies into keras-js rather keras-js initializing them on its own.





On Tue, Nov 29, 2016 at 9:13 PM -0800, "Leon Chen" <notifications@github.com> wrote:












I've tried it -- can definitely make this work in nodejs after stripping away weblas/canvas/webgl. What's your compelling reason for wanting to run it in nodejs, though?

**Comment by nakosung** [comment]
it would be convenient if keras.js is provided as a npm module. (like convnetjs)
```
Install it: $ npm install convnetjs
Use it: var convnetjs = require("convnetjs");
```

npm module is not mean to be used only within node.js. It can be packed within a web app or other exciting platforms, like Atom, VSCode and [unreal.js](https://github.com/ncsoft/Unreal.js).

**Comment by transcranial** [comment]
That's pretty interesting @nakosung . Don't worry, this is definitely coming.

**Comment by kevinmartinmauro** [comment]
I think this  would likely fix #40 for my use case too

**Comment by nemo** [comment]
@transcranial - is this possible already?

Want the same use-case of having models built with Keras/TF ahead of time and using node for the backend (since we're already using it)

**Comment by zixia** [comment]
+1

**Comment by transcranial** [comment]
Finally got around to adding node support -- added some notes in the README so you guys can get started. It's working fine in my preliminary tests.
